app/models/job_execution.rb

```diff
@@ -142,6 +132,25 @@ def setup!(dir)
     end
   end
 
+  def commands(dir)
+    @commands ||= begin
+      commands = [
+        "export DEPLOYER=#{@job.user.email.shellescape}",
+        "export DEPLOYER_EMAIL=#{@job.user.email.shellescape}",
+        "export DEPLOYER_NAME=#{@job.user.name.shellescape}",
+        "export REVISION=#{@reference.shellescape}",
+        "export TAG=#{(@job.tag || @job.commit).to_s.shellescape}",
+        "export CACHE_DIR=#{artifact_cache_dir}",
+        "cd #{dir}",
+        *@job.commands
+      ]
+      if @stage && @stage.deploy_groups.count > 0
```

Comment:
saving an extra query and map! for a bit of performance win:
`if @stage && group_names = @stage.deploy_groups.pluck(:name).map!(&:shellescape).join(" ")`